packages/vm/src/imports.rs

```diff
@@ -253,45 +234,19 @@ pub fn do_scan<S: Storage + 'static, Q: Querier>(
     start_ptr: u32,
     end_ptr: u32,
     order: i32,
-) -> i32 {
-    let start = match maybe_read_region(ctx, start_ptr, MAX_LENGTH_DB_KEY) {
-        Ok(data) => data,
-        Err(VmError::RegionLengthTooBigErr { .. }) => return errors::REGION_READ_LENGTH_TOO_BIG,
-        Err(_) => return errors::REGION_READ_UNKNOWN,
-    };
-    let end = match maybe_read_region(ctx, end_ptr, MAX_LENGTH_DB_KEY) {
-        Ok(data) => data,
-        Err(VmError::RegionLengthTooBigErr { .. }) => return errors::REGION_READ_LENGTH_TOO_BIG,
-        Err(_) => return errors::REGION_READ_UNKNOWN,
-    };
+) -> VmResult<i32> {
+    let start = maybe_read_region!(ctx, start_ptr, MAX_LENGTH_DB_KEY);
+    let end = maybe_read_region!(ctx, end_ptr, MAX_LENGTH_DB_KEY);
     let order: Order = match order.try_into() {
-        Ok(o) => o,
-        Err(_) => return errors::scan::INVALID_ORDER,
+        Ok(order) => order,
+        Err(_) => return Ok(errors::scan::INVALID_ORDER),
     };
-    let range_result = with_storage_from_context::<S, Q, _, _>(ctx, |store| {
-        let iter = match store.range(start.as_deref(), end.as_deref(), order) {
-            Ok(iter) => iter,
-            Err(_) => return make_runtime_err("An error occurred in range call"),
-        };
+    let iterator = with_storage_from_context::<S, Q, _, _>(ctx, |store| {
+        Ok(store.range(start.as_deref(), end.as_deref(), order)?)
+    })?;
 
-        // Unsafe: I know the iterator will be deallocated before the storage as I control the lifetime below
-        // But there is no way for the compiler to know. So... let's just lie to the compiler a little bit.
-        let live_forever: Box<dyn Iterator<Item = StdResult<KV>> + 'static> =
-            unsafe { mem::transmute(iter) };
-        Ok(live_forever)
-    });
```

Comment:
@ethanfrey @webmaster128 
A lot of the changes made in `context.rs` (and elsewhere) around lifetimes has been done in order to prove that this unsafe block is unnecessary. I'm pretty sure it had potential to invoke UB. You can see the core problem described in the big comment in `context.rs::get_context_data`. If necessary i can explain things in more detail but there's a fundamental issue where `add_iterator` below tried to mutate `ctx` while it is mutably borrowed by the iterator we're trying to register to it. The solution is to find the right place to dissociate the lifetimes of all these components, but to do that we'll need some `Rc<RefCell<T>>` somewhere, or some other indirection mechanism...

Reply:
please respond to this, I'm not sure it's wise to keep it this way long term but i'm also not sure what's the best path forward with it.

Reply:
Where is the iterator holding a reference to the storage? In this Go code?

```go
//export cScan
func cScan(ptr *C.db_t, start C.Buffer, end C.Buffer, order i32, out *C.GoIter) (ret C.GoResult) {
	defer recoverPanic(&ret)

	kv := *(*KVStore)(unsafe.Pointer(ptr))
	// handle null as well as data
	var s, e []byte
	if start.ptr != nil {
		s = receiveSlice(start)
	}
	if end.ptr != nil {
		e = receiveSlice(end)
	}

	var iter dbm.Iterator
	switch order {
	case 1: // Ascending
		iter = kv.Iterator(s, e)
	case 2: // Descending
		iter = kv.ReverseIterator(s, e)
	default:
		return C.GoResult_BadArgument
	}
	// Let's hope this works!
	*out = buildIterator(iter)
	return C.GoResult_Ok
}
```

Reply:
Yeah, in:
```rust
iter = kv.Iterator(s, e)
```
The resulting iterator references the db, which is named `kv`
It's the same case for the mock storage, where the iterator references a hashmap, and this is also reflected in the type signature of the `Storage` trait.

Reply:
@reuvenpo I responded below in some comments, but I did fight with this issue a fair bit.

Attempt one had all find of lifetime issues, and I just eagerly read all iterator results into a slice and returned a reference to this slice, so the iterator no longer had any reference to the backing storage. This fixed all lifetime issues (no hanging references), but is super inefficient, especially when the contract just wants the first/last item in a range.

I then realized I can add a lifetime to the box and got this working *inside cosmwasm_std* with lazy iterators and the reference of the iterator to the storage, with the compiler-defined guarantee that the iterator would be dropped before the storage is dropped.

I spent some time in context to try to move them together or such to let the compile understand these lifetimes and accept them as safe, but that was many hours in vain. In the end, I found `transmute` and used it for a little lie to the compiler. But saying "everything will be alright". Like I say below, if you can guarantee that all `iterator` references will be released before the `storage` one (in context data), then I am correctly using `unsafe` and not introducing `UB`. If you find a way to drop storage without dropping the iterators first, then point it out and let's fix it.

I think best to analyze this in the pure Rust case first, we do the same lifetime issues using MockStorage, and this is easier to analyze. The Go case is the same issue, but one more level of indirection.

Reply:
Sounds right. Let's do it in a separate PR, because i'd rather have it as a focused task

Reply:
>Like I say below, if you can guarantee that all iterator references will be released before the storage one (in context data), then I am correctly using unsafe and not introducing UB

unfortunately, that's not how UB works in rust. If you write code in a way that implies things are true, then the compiler has full rights to assume that they are, which causes incorrect code generation. so for example with the double mutation of the context that i described, the compiler can assume that it isn't being changed between two reads, when in reality we do change things that are "owned" by the context. In the FFI case it can't do it, but in the tests which back the storage with a plain Rust HashMap, we might be breaking the internal state of the hashmap, just by doing that one trick.

>If you find a way to drop storage without dropping the iterators first, then point it out and let's fix it

With my code as it is today you can't do that, but you can try to use the ContextData after the context destructor has cleared it (or worse, while the context destructor is cleaning it)

Reply:
Resolving in favour of https://github.com/CosmWasm/cosmwasm/issues/335